app/components/app_activity_log_component.rb

```diff
@@ -318,59 +318,93 @@ def programmes_by_id
+            patient.vaccination_status(programme:, academic_year:).vaccinated?
+          end
+
+        vaccinated_but_seasonal_programmes =
+          all_programmes.select do |programme|
+            patient.vaccination_status(
+              programme:,
+              academic_year:
+            ).vaccinated? && programme.seasonal?
+          end
+
+        expired_items =
+          {
+            vaccinated_but_seasonal: vaccinated_but_seasonal_programmes,
+            not_vaccinated: not_vaccinated_programmes
+          }.transform_values do |programmes|
+            get_expired_items(academic_year:, programmes:)
+          end
+
+        expired_items.map do |category, expired_items_in_category|
+          expired_item_names = []
+          if expired_items_in_category[:consents].any?
+            expired_item_names += ["consent", "health information"]
+          end
+          if expired_items_in_category[:triages].any?
+            expired_item_names << "triage outcome"
+          end
+          if expired_items_in_category[:psds].any?
+            expired_item_names << "PSD status"
+          end
+
+          next if expired_item_names.empty?
 
-      programmes_with_expired_consents =
-        consents
-          .select { it.academic_year == academic_year }
-          .flat_map { programmes_for(it) }
-          .reject { vaccinated_programmes.include?(it) }
-
-      programmes_with_expired_triages =
-        triages
-          .select { it.academic_year == academic_year }
-          .flat_map { programmes_for(it) }
-          .reject { vaccinated_programmes.include?(it) }
-
-      programmes_with_expired_psds =
-        patient_specific_directions
-          .select { it.academic_year == academic_year }
-          .flat_map { programmes_for(it) }
-          .reject { vaccinated_programmes.include?(it) }
-
-      expired_items = []
-      if programmes_with_expired_consents.any?
-        expired_items += ["consent", "health information"]
+          title =
+            "#{
+              expired_item_names.to_sentence(
+                words_connector: ", ",
+                last_word_connector: " and "
+              ).upcase_first
+            } expired"
+
+          body =
+            case category
+            when :not_vaccinated
+              "#{@patient.full_name} was not vaccinated."
+            when :vaccinated_but_seasonal
+              "#{@patient.full_name} was vaccinated."
+            end
+
+          programmes = expired_items_in_category.values.flatten.uniq
+
+          {
+            title:,
+            body:,
+            at:
+              academic_year.to_academic_year_date_range.end.end_of_day -
+                1.second,
+            programmes:
+          }
+        end
       end
-      expired_items << "triage outcome" if programmes_with_expired_triages.any?
-      expired_items << "PSD status" if programmes_with_expired_psds.any?
 
-      next if expired_items.empty?
+    result.compact
```

Comment:
If we do `next []` inside the block, we don't need to worry about calling `compact`, it also means the linter won't complain about calling `.compact` and `end` 😆